indexer/services/ender/src/lib/kafka-publisher.ts

```diff
@@ -43,34 +55,135 @@ export class KafkaPublisher {
+   *
+   * Due to separate handlers for order fills, we can be sure that if a message is annotated
+   * with a fill, it should only contain data about a fill / order and not transfers.
    */
-  public sortTradeEvents() {
-    this.tradeMessages = this.tradeMessages.sort(
-      (a: SingleTradeMessage, b: SingleTradeMessage) => {
+  public aggregateFillEventsForSubaccountMessages() {
+    // Create a map to store the last event for fills per order ID
+    const lastEventForFills: Record<string, AnnotatedSubaccountMessage> = {};
+    // Create a map to store all fill events per order ID
+    const allFillEvents: Record<string, FillSubaccountMessageContents[]> = {};
+    const nonFillEvents: AnnotatedSubaccountMessage[] = [];
+
+    this.subaccountMessages.forEach((message: AnnotatedSubaccountMessage) => {
+      if (message.isFill && message.orderId) {
+        const fills:
+        FillSubaccountMessageContents[] | undefined = this.extractFills(message.contents);
+        const orderId = message.orderId;
+        if (fills !== undefined) {
+          allFillEvents[orderId] = allFillEvents[orderId]
+            ? allFillEvents[orderId].concat(fills)
+            : fills;
+        }
+
+        // If we haven't seen this order ID before or if the current message
+        // was associated with a later event, update the lastFillEvents for this order ID
+        if (
+          !lastEventForFills[orderId] ||
+          this.compareMessages(message, lastEventForFills[orderId]) > 0
+        ) {
+          lastEventForFills[orderId] = message;
+        }
+      } else {
+        nonFillEvents.push(message);
+      }
+    });
+
+    // Replace the fills in the last event for fills with all the fills for that order ID
+    Object.keys(lastEventForFills).forEach((orderId: string) => {
+      const lastEvent = lastEventForFills[orderId];
+      const fills = allFillEvents[orderId];
+      if (fills) {
+        lastEvent.contents = this.replaceFills(lastEvent.contents, fills);
+      }
+    });
+
+    this.subaccountMessages = Object.values(lastEventForFills)
+      .concat(nonFillEvents)
+      .map((annotatedMessage) => convertToSubaccountMessage(annotatedMessage));
+    this.sortEvents(this.subaccountMessages);
+  }
+
+  /** Helper function to compare two AnnotatedSubaccountMessages based on block height,
+   * transaction index, and event index.
+   */
+  private compareMessages(a: AnnotatedSubaccountMessage, b: AnnotatedSubaccountMessage) {
+    if (a.blockHeight === b.blockHeight) {
+      if (a.transactionIndex === b.transactionIndex) {
+        return a.eventIndex - b.eventIndex;
+      }
+      return a.transactionIndex - b.transactionIndex;
+    }
+    return Number(a.blockHeight) - Number(b.blockHeight);
+  }
+
+  /**
+   * Sort events by block height, transaction index, and event index in ascending order,
+   * where the first event should be the earliest event in the block.
+   */
+  public sortEvents(msgs: OrderedMessage[]) {
+    if (msgs) {
```

Comment:
Is this needed? Would calling `.sort` on an empty array lead to an error / undesirable outcome?

Reply:
removed